bluebird/bluebird-tests.ts

```diff
@@ -236,36 +253,96 @@ barProm = barProm.then((value: Bar) => {
 
 // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 
-barProm = fooProm.catch((reason: any) => {
+fooProm = fooProm.catch((reason: any) => {
+	return;
+});
+
+fooProm = fooProm.caught((reason: any) => {
+	return;
+});
+fooProm = fooProm.catch((error: any) => {
+	return true;
+}, (reason: any) => {
+	return;
+});
+fooProm = fooProm.caught((error: any) => {
+	return true;
+}, (reason: any) => {
+	return;
+});
+
+fooProm = fooProm.catch((reason: any) => {
+	return voidProm;
+});
+
+fooProm = fooProm.caught((reason: any) => {
+	return voidProm;
+});
+fooProm = fooProm.catch((error: any) => {
+	return true;
+}, (reason: any) => {
+	return voidProm;
+});
+fooProm = fooProm.caught((error: any) => {
+	return true;
+}, (reason: any) => {
+	return voidProm;
+});
+
+fooProm = fooProm.catch((reason: any) => {
+	//handle multiple valid return types simultaneously
+	if (true) {
+		return;
+	} else if (false) {
+		return voidProm;
+	} else if (foo) {
+		return foo;
+	}
+});
+
+fooOrBarProm = fooProm.catch((reason: any) => {
 	return bar;
 });
-barProm = fooProm.caught((reason: any) => {
+fooOrBarProm = fooProm.caught((reason: any) => {
```

Comment:
I'm not sure this is important behavior, it certainly seems like a bad practice.  But it's definitely possible with bluebird and the previous tests were all focused on changing the type of the promise.catch so I added the union.